### ISSUE TITLE
C-a and C-e  going to the begging and ending of line instead of default

### ISSUE THREAD

**Issue by 3ximus** [akinomyoga/ble.sh#83]
Hi, 2 issues in a row this time xD
I wanted to report this a while ago but never got around to it.

By default C-a and C-e in bash will move the cursor to the beginning and ending of a command, but with ble.sh they move to the beginning of the line (in case of a command that spans more than 1 line). I prefer the default behavior of jumping to the beginning of the command, how can I bind it to have that behavior?

![jumping](https://user-images.githubusercontent.com/9083012/105387769-c8e96000-5c0d-11eb-8ff3-94bc099c3f7a.gif)

Thank you again for this project. I wish I could help instead of just sending you issues, but I don't really know where to start and I don't wan't to fuck your codebase xD

Cheers!!

**Comment by akinomyoga** [comment]
Thank you! This is actually configurable though I haven't described it in the manual.

```bash
ble-bind -m 'emacs' -f 'home' '@nomarked beginning-of-logical-line'
ble-bind -m 'emacs' -f 'C-a' '@nomarked beginning-of-logical-line'
ble-bind -m 'emacs' -f 'end' '@nomarked end-of-logical-line'
ble-bind -m 'emacs' -f 'C-e' '@nomarked end-of-logical-line'
ble-bind -m 'emacs' -f 'S-home' '@marked beginning-of-logical-line'
ble-bind -m 'emacs' -f 'C-S-a' '@marked beginning-of-logical-line'
ble-bind -m 'emacs' -f 'S-end' '@marked end-of-logical-line'
ble-bind -m 'emacs' -f 'C-S-e' '@marked end-of-logical-line'
ble-bind -m 'emacs' -f 'C-k' 'kill-forward-logical-line'
ble-bind -m 'emacs' -f 'C-u' 'kill-backward-logical-line'
```

Maybe I'll later add a single option to control the behavior such like `bleopt line_type=graphical/logical`

**Comment by 3ximus** [comment]
Ok nice, thanks !

Btw, what about the behavior of 'C-w' ? How do I set it to default? I tried 'kill-word' and  'backward-kill-word' but none of them seem to work

**Comment by akinomyoga** [comment]
> Btw, what about the behavior of 'C-w' ? How do I set it to default? I tried 'kill-word' and 'backward-kill-word' but none of them seem to work

I assume you are using emacs mode (but not vi mode). If that's not the case, please let me know. It depends on what behavior you expect in detail, but there are several types of word boundaries in `ble.sh`: `cword`, `eword`, `fword`, `sword`, and `uword`.

- `eword` is a sequence of ASCII alphanumeric letters.
- `cword` is a sequence of ASCII alphanumeric letters plus underscores (`_`).
- `uword` is a sequence of non-space characters.
- `fword` is a sequence of non-space characters excluding slashes (`/`).
- `sword` is a sequence of characters that are not the shell special characters (spaces, `|&|<>()`)

Possible settings are

```bash
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-eword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-cword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-uword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-fword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-sword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-eword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-cword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-uword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-fword'
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-sword'
```

I think the setting closest to the normal Bash is

```bash
ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-uword'
```

Exactly speaking, just `kill-backward-uword` (without `kill-region-or `) would be the closest to the normal Bash, but I recommend including `kill-region-or ` which switches the behavior to cut the currently selected text when there is a selection.

If you still see some differences in detailed behavior, please feel free to describe them here.

**Comment by 3ximus** [comment]
> ```shell
> ble-bind -m 'emacs' -f 'C-w' 'kill-region-or kill-backward-uword'
> ```

Yup, this works fine like i'm used to. Thank you !! 
There is another difference I noticed... Multiple `kill` comands append to the bash readline clipboard, so you can do `C-w` and `C-u` and if they are successive key presses they append to this clipboard. Sometimes its useful if for example you want to delete the last 2 words you just typed to paste later or or somewhere else on the line.

**Comment by akinomyoga** [comment]
> There is another difference I noticed... Multiple kill comands append to the bash readline clipboard, so you can do C-w and C-u and if they are successive key presses they append to this clipboard. Sometimes its useful if for example you want to delete the last 2 words you just typed to paste later or or somewhere else on the line.

Oh, OK. That's a known difference which I personally don't use so much. I'll support it after I investigate the detailed behavior of Readline and GNU Emacs (where the feature originally came from). Thank you for reminding me of that!

**Comment by 3ximus** [comment]
Yeah, don't worry about it, it's not that big of a deal.
I'll close this issue since my main question is resolved.
Thank you once again !!

**Comment by akinomyoga** [comment]
Ah, OK. I was thinking something like "I'm going to support this feature *some day*" and didn't support it for a long time. Actually, there are many such features, and combining killed strings is one of them. I think this is a good chance to support it, so anyway I'll consider it. Thank you!

For the behavior of the beginning and the end of lines, in the commit 40ae242,
- I added an option [`bleopt edit_line_type`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_line_type)
- and also changed the default behavior to use the logical line (`bleopt edit_line_type=logical`).

I was thinking to finish both of them (supports for `bleopt edit_line_type` and combining killed strings) and push them at once, but now I decided to push `bleopt edit_line_type` first.

For the discussion in the other issue #82, could you give me some time to think about it in detail?

**Comment by 3ximus** [comment]
> For the behavior of the beginning and the end of lines, in the commit [40ae242](https://github.com/akinomyoga/ble.sh/commit/40ae242754d7606c268da89168b0e84b17c06da6),
> 
>     * I added an option [`bleopt edit_line_type`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_line_type)
> 
>     * and also changed the default behavior to use the logical line (`bleopt edit_line_type=logical`).

Nice! Perfect! thank you!

> For the discussion in the other issue #82, could you give me some time to think about it in detail?

Yeah, of course!!
Cheers !!

**Comment by akinomyoga** [comment]
66564e1 I now supported combining kills/copies. Thank you!

**Comment by 3ximus** [comment]
Hi, I think you might have missed something, maybe resetting the copied content correctly:

![combined kills](https://user-images.githubusercontent.com/9083012/105714659-8927be80-5f14-11eb-8366-ea554a1415d9.gif)

Right now it seems that it's saving too much, and this makes C-y unusable also since it still has previous content saved.
Can you reproduce this behavior?

**Comment by akinomyoga** [comment]
Oh, thank you for testing, and sorry for such a trivial bug. This was caused by the widget `kill-region-or`. In the previous commit, I tested with `kill-backward-uword` but forgot to test with the combination `kill-region-or kill-backward-uword`. Now I fixed it in the commit 1631751.

In the commit 47a3301, I also decided to change the default behavior of <kbd>C-w</kbd> in `ble.sh` as suggested by you in https://github.com/akinomyoga/ble.sh/issues/83#issuecomment-764893198. I also changed <kbd>M-w</kbd> similarly.

<details><summary>Background of changing the key binding to C-w</summary>

I tested *combining kills* with the old binding `kill-region-or kill-uword` and found that it behaves strangely in combination with this *combining kills*. I traced back the change history of <kbd>C-w</kbd> in `ble.sh` to find that it actually unchanged from the very beginning commit of `ble.sh` when `ble.sh` was just an experimental implementation.

```
# Change history of C-w
bbbd155f src/edit.sh (Koichi Murase 2019-03-22 07:28:24 +0900 6763)   ble-decode/keymap:safe/.bind 'C-w'       'kill-region-or kill-uword'
1fc7cbaf ble-edit.sh (Koichi Murase 2017-12-04 20:48:17 +0900 6185)   ble-decode/keymap:safe/.bind 'C-w'       'kill-region-or uword'
f18485f0 ble-edit.sh (Koichi Murase 2017-12-04 14:36:52 +0900 4442)   ble-bind -f 'C-w'       'kill-region-or uword'
3aa7fa66 ble-edit.sh (Koichi Murase 2017-12-03 18:31:00 +0900 4461)   ble-bind -f 'C-w'      'kill-region-or uword'
6ca737d2 keymap/emacs.sh (Koichi Murase 2015-02-28 12:48:55 +0900  54)   ble-bind -f 'C-w'      'kill-region-or uword'
^c68412b ble-edit.sh (Koichi Murase 2015-02-09 03:13:19 +0900 3381)   ble-bind -f 'C-w'      'kill-region-or uword'
```

I think there is really no reason to do `kill-region-or kill-uword` by <kbd>C-w</kbd> so decided to change it to match with Readline.
</details>

Thank you for the report!

**Comment by 3ximus** [comment]
Yeah, well I still prefer the default behavior of readline since word kill works the same in most applications, so it's easier if it has the same behavior everywhere.
Thank you for your updates!
Cheers!
